kura/org.eclipse.kura.cloudconnection.sparkplug.mqtt.provider/src/main/java/org/eclipse/kura/cloudconnection/sparkplug/mqtt/device/SparkplugDevice.java

```diff
@@ -0,0 +1,224 @@
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.cloudconnection.sparkplug.mqtt.device;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.eclipse.kura.KuraErrorCode;
+import org.eclipse.kura.KuraException;
+import org.eclipse.kura.cloudconnection.CloudConnectionConstants;
+import org.eclipse.kura.cloudconnection.CloudConnectionManager;
+import org.eclipse.kura.cloudconnection.listener.CloudConnectionListener;
+import org.eclipse.kura.cloudconnection.listener.CloudDeliveryListener;
+import org.eclipse.kura.cloudconnection.message.KuraMessage;
+import org.eclipse.kura.cloudconnection.publisher.CloudPublisher;
+import org.eclipse.kura.cloudconnection.sparkplug.mqtt.endpoint.SparkplugCloudEndpoint;
+import org.eclipse.kura.cloudconnection.sparkplug.mqtt.message.SparkplugMessageType;
+import org.eclipse.kura.configuration.ConfigurableComponent;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.Filter;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SparkplugDevice
+        implements CloudPublisher, ConfigurableComponent, CloudConnectionListener, CloudDeliveryListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(SparkplugDevice.class);
+
+    public static final String KEY_MESSAGE_TYPE = "message.type";
+    public static final String KEY_DEVICE_ID = "device.id";
+
+    private String deviceId;
+    private BundleContext bundleContext;
+    private ServiceTracker<CloudConnectionManager, CloudConnectionManager> cloudConnectionManagerTracker;
+    private Optional<SparkplugCloudEndpoint> sparkplugCloudEndpoint = Optional.empty();
+    private final Set<CloudConnectionListener> cloudConnectionListeners = new CopyOnWriteArraySet<>();
+    private final Set<CloudDeliveryListener> cloudDeliveryListeners = new CopyOnWriteArraySet<>();
+    private final ExecutorService callbackDispatcher = Executors.newCachedThreadPool();
+    private Set<String> deviceMetrics = new HashSet<>();
+
+    /*
+     * ConfigurableComponent APIs
+     */
+
+    public void activate(final ComponentContext componentContext, final Map<String, Object> properties)
+            throws InvalidSyntaxException {
+        this.bundleContext = componentContext.getBundleContext();
+
+        String selectedCloudEndpointPid = (String) properties
```

Comment:
Not sure if we are doing the same in other cloud connectors, but since the tracker startup is a lengthy process, does it make sense to have it in the activate or better in a separate thread?

Reply:
Agree